### PR TITLE
Refactor PT 3.x  API: static_quant & rtn

### PR DESCRIPTION
**Pull request by yuwenzho**
## Type of Change

feature: Refactor PT 3.x static quant API
API changed or not: yes

## Description

Refactor PT 3.x static quant API

- Add new user-facing API: `from neural_compressor.torch.quantization import prepare` and `from 
 neural_compressor.torch.quantization import convert`
- Add base class: `Quantizer`. Any algorithm needs to inherit `Quantizer` base class.
- Modify internal algorithm entry to call member functions of algorithm Quantizer class.

### Usage
**static quant**
```diff
- from neural_compressor.torch.quantization import get_default_static_config, quantize
+ from neural_compressor.torch.quantization import get_default_static_config, prepare, convert

quant_config = get_default_static_config()
example_inputs = self.input
- q_model = quantize(fp32_model, quant_config=quant_config, run_fn=run_fn, example_inputs=example_inputs)
+ prepared_model = prepare(fp32_model, quant_config=quant_config, example_inputs=example_inputs)
+ run_fn(prepared_model)
+ q_model = convert(prepared_model)
```

**rtn quant** (user-facing API will not change)
```python
from neural_compressor.torch.quantization import RTNConfig, quantize
quant_config = RTNConfig()
q_model = quantize(model, quant_config)
```
## How has this PR been tested?

CI

## Dependency Change?

No

### CODE REVIEW

**Comment by xin3he** [comment]
@yuwenzho, @violetch24 ,  You may have conflict with each other, please have a sync before merge.

**Comment by yuwenzho** [comment]
> @yuwenzho, @violetch24 , You may have conflict with each other, please have a sync before merge. #1740

will merge PR  #1740 first and I will rebase it later.

File: neural_compressor/torch/algorithms/base_algorithm.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+def algo_quantizer_register(name):
+    def decorator(algo_quantizer):
+        algo_quantizers[name] = algo_quantizer
+        return algo_quantizer
+
+    return decorator
+
+
+class BaseQuantizer(ABC):
```
Comment (yiliu30) [inline neural_compressor/torch/algorithms/base_algorithm.py:28]: Suggest to follow this interface design. https://github.com/pytorch-labs/ao/blob/3124382b424beac3d7671beba0ea65a51daf137c/torchao/quantization/unified.py#L15

File: neural_compressor/torch/algorithms/base_algorithm.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import ABC, abstractmethod
+
+algo_quantizers = {}
+
+
+def algo_quantizer_register(name):
```
Comment (yiliu30) [inline neural_compressor/torch/algorithms/base_algorithm.py:20]: nit: we need to refine the function naming later.

File: neural_compressor/torch/algorithms/weight_only/__init__.py [hunk 3]
```diff
@@ -12,7 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from .rtn import rtn_quantize
+from .rtn import RTNQuantizer
```
Comment (yiliu30) [inline neural_compressor/torch/algorithms/weight_only/__init__.py:15]: This file has been emptied by https://github.com/intel/neural-compressor/pull/1742, please rebase the master.

**Review by yiliu30 — APPROVED** [review]

File: neural_compressor/torch/algorithms/weight_only/awq.py [hunk 4]
```diff
@@ -429,14 +429,15 @@ def apply_quantize_with_clip(self, return_int=False):
...
-        self.model = rtn_quantize(
+        rtn_quantizer = RTNQuantizer()
+        rtn_quantizer.config = self.weight_config
+        rtn_quantizer.quantize(
             self.model,
-            num_bits=self.bits,
+            bits=self.bits,
             group_size=self.group_size,
             scheme=self.scheme,
-            weight_config=self.weight_config,
```
Comment (xin3he) [inline neural_compressor/torch/algorithms/weight_only/awq.py:439]: This is not correct, AWQ is using weight_config to notice RTN how to quantize with clip range.

File: neural_compressor/torch/algorithms/weight_only/rtn.py [hunk 5]
```diff
@@ -19,174 +19,209 @@
...
-    use_full_range=False,
-    use_mse_search=False,
-    **kwargs,
-):
-    """Quant the model with round to nearest method and inplace is True.
+@algo_quantizer_register(name=RTN)
+class RTNQuantizer(Quantizer):
+    def __init__(self, config: OrderedDict = {}):
+        super().__init__(config)
+        self._preprocess_config()
```
Comment (xin3he) [inline neural_compressor/torch/algorithms/weight_only/rtn.py:37]: @yiliu30 I assume that this change aims to allow user recover the quantization with config_mapping recovered from json file, or I don't know how user can get the config_mapping that INC designed.

File: neural_compressor/torch/quantization/algorithm_entry.py [hunk 6]
```diff
@@ -40,32 +40,12 @@ def rtn_entry(
...
-            "double_quant_dtype": quant_config.double_quant_dtype,
-            "double_quant_bits": quant_config.double_quant_bits,
-            "double_quant_scheme": "sym" if quant_config.double_quant_use_sym else "asym",
-            "double_quant_group_size": quant_config.double_quant_group_size,
-        }
-
-    model = rtn_quantize(model, weight_config=weight_config)
-    return model
+    logger.info("Quantize model with the RTN algorithm.")
+    algo = RTNQuantizer(configs_mapping)
```
Comment (xin3he) [inline neural_compressor/torch/quantization/algorithm_entry.py:46]: Here we need to clarify whether the algorithm should allow users to construct the "config_mapping" designed by INC or allow users to construct a simple dictionary. @ftian1

File: neural_compressor/torch/algorithms/base_algorithm.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    def __init__(self, config: OrderedDict = {}):
+        self.config = config
+
+    @abstractmethod
+    def prepare(self, model: torch.nn.Module, *args: Any, **kwargs: Any) -> torch.nn.Module:
+        raise NotImplementedError
+
+    @abstractmethod
+    def convert(self, model: torch.nn.Module, *args: Any, **kwargs: Any) -> torch.nn.Module:
+        raise NotImplementedError
```
Comment (xin3he) [inline neural_compressor/torch/algorithms/base_algorithm.py:55]: Since one-step and two-step use different function names, is it reasonable for us to merge these two quantizers together? @yiliu30

**Review by xin3he — CHANGES_REQUESTED** [review]

**Comment by xin3he** [comment]
@Kaihui-intel, Please aware these changes will impact what you are doing.

File: neural_compressor/torch/algorithms/weight_only/awq.py (hunk 4 above)
Comment (yuwenzho) [reply neural_compressor/torch/algorithms/weight_only/awq.py:439]: weight_config here is set as a member variables of RTNQuantizer. You can see it on Line-435 `rtn_quantizer.config = self.weight_config`

File: neural_compressor/torch/quantization/quantize.py [hunk 8]
```diff
@@ -81,5 +82,124 @@ def quantize(
                 run_fn=run_fn,
                 run_args=run_args,
                 example_inputs=example_inputs,
+                mode="quantize",
+            )
+    return q_model
+
+
+@log_quant_execution
```
Comment (yiliu30) [inline neural_compressor/torch/quantization/quantize.py:90]: `log_quant_execution` prints something like `Quantization started`, `Quantization ended`. Suggest removing it first and making the `log_quant_execution` more general in a separate PR.

**Review by yiliu30 — CHANGES_REQUESTED** [review]

File: neural_compressor/torch/algorithms/weight_only/rtn.py (hunk 5 above)
Comment (yuwenzho) [reply neural_compressor/torch/algorithms/weight_only/rtn.py]: preprocess config step was moved back to algorithm_entry.py

File: neural_compressor/torch/algorithms/base_algorithm.py (hunk 7 above)
Comment (yuwenzho) [reply neural_compressor/torch/algorithms/base_algorithm.py]: The two Quantizer calsses was merged together.

File: neural_compressor/torch/algorithms/base_algorithm.py [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+    def __init__(self, config: OrderedDict = {}):
+        self.config = config
+
+    @abstractmethod
+    def quantize(self, model: torch.nn.Module, *args: Any, **kwargs: Any) -> torch.nn.Module:
+        raise NotImplementedError
+
+
+class TwoStepQuantizer(ABC):
```
[the remark this answers is not in the document]
Comment (yuwenzho) [reply neural_compressor/torch/algorithms/base_algorithm.py]: New Quantizer class no longer originates from torchao

File: neural_compressor/torch/quantization/quantize.py (hunk 8 above)
Comment (yuwenzho) [reply neural_compressor/torch/quantization/quantize.py]: removed

File: neural_compressor/torch/algorithms/base_algorithm.py [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    def convert(self, model: torch.nn.Module, *args: Any, **kwargs: Any) -> torch.nn.Module:
+        raise NotImplementedError(
+            "{} doesn't implement `convert` function. "
+            "Please implement `convert` function or "
+            "switch to call `from neural_compressor.torch.quantization import quantize` to do quantization.".format(
+                self.__class__.__name__
+            )
+        )
+
+    def execute(self, model, mode, *args: Any, **kwargs: Any):
```
[the remark this answers is not in the document]
Comment (yuwenzho) [reply neural_compressor/torch/algorithms/base_algorithm.py:45]: updated

File: neural_compressor/torch/algorithms/base_algorithm.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import ABC, abstractmethod
+from collections import OrderedDict
+from typing import Any
+
+import torch
+
+algo_quantizers = {}
```
[the remark this answers is not in the document]
Comment (yuwenzho) [reply neural_compressor/torch/algorithms/base_algorithm.py]: removed

File: neural_compressor/torch/algorithms/base_algorithm.py [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+from collections import OrderedDict
+from typing import Any
+
+import torch
+
+from neural_compressor.torch.utils import Mode
+
+
+class Quantizer(ABC):
+    """The base quantizer for one step algorithm quantizers."""
```
Comment (yiliu30) [inline neural_compressor/torch/algorithms/base_algorithm.py:25]: Please update the docstirng.

File: neural_compressor/torch/quantization/algorithm_entry.py [hunk 13]
```diff
@@ -64,7 +64,10 @@ def rtn_entry(
             "double_quant_group_size": quant_config.double_quant_group_size,
         }
 
-    model = rtn_quantize(model, weight_config=weight_config)
+    mode = kwargs.get("mode", "quantize")
```
Comment (yiliu30) [inline neural_compressor/torch/quantization/algorithm_entry.py:67]: Recommend making the `mode` as a mandatory argument for the entry function.

**Review by yiliu30 — APPROVED** [review]

File: test/3x/torch/quantization/test_static_quant.py [hunk 14]
```diff
@@ -152,3 +158,137 @@ def run_fn(model):
...
+        fp32_model.linear.weight = torch.nn.Parameter(torch.tensor([[0.0, 1.0], [1.0, 0.0]]))
+        example_inputs = torch.zeros(3, 2)
+        quant_config = StaticQuantConfig(act_sym=True, act_algo="kl")
+        prepared_model = prepare(fp32_model, quant_config=quant_config, example_inputs=example_inputs)
+        run_fn(prepared_model)
+        q_model = convert(prepared_model)
+        output1 = fp32_model(example_inputs)
+        output2 = q_model(example_inputs)
+        # set a big atol to avoid random issue
+        assert torch.allclose(output1, output2, atol=2e-2), "Accuracy gap atol > 0.02 is unexpected. Please check."
```
Comment (Kaihui-intel) [inline test/3x/torch/quantization/test_static_quant.py:246]: can you add an `assert` that meets the accuracy gap?

File: neural_compressor/torch/algorithms/weight_only/rtn.py [hunk 15]
```diff
@@ -19,174 +19,182 @@
...
-    quantile=1.0,
-    weight_config={},
-    export_compressed_model=False,
-    use_full_range=False,
-    use_mse_search=False,
-    **kwargs,
-):
-    """Quant the model with round to nearest method and inplace is True.
+class RTNQuantizer(Quantizer):
+    def __init__(self, tune_cfg: OrderedDict = {}):
```
Comment (Kaihui-intel) [inline neural_compressor/torch/algorithms/weight_only/rtn.py:33]: Please add docstring about tune_cfg.

File: neural_compressor/torch/algorithms/base_algorithm.py (hunk 12 above)
Comment (yuwenzho) [reply neural_compressor/torch/algorithms/base_algorithm.py]: updated

File: neural_compressor/torch/quantization/algorithm_entry.py (hunk 13 above)
Comment (yuwenzho) [reply neural_compressor/torch/quantization/algorithm_entry.py]: updated

File: neural_compressor/torch/algorithms/weight_only/rtn.py (hunk 15 above)
Comment (yuwenzho) [reply neural_compressor/torch/algorithms/weight_only/rtn.py:33]: updated

**Review by xin3he — APPROVED** [review]

**Comment by yuwenzho** [comment]
> is `example_inputs` mandatory?

Not mandatory, `example_inputs` is required for static quantize.

**Review by ftian1 — APPROVED** [review]
